### PR TITLE
Adds mapping for Failed pod phase

### PR DESCRIPTION
**Pull request by dposada**
## Changes proposed in this PR

- adding handling to `pod->synthesized-pod-state` for the `Failed` [pod phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase)
- adding handling to `container-status->failure-reason` for the case where the pod failed, there is no container status, and the pod status reason is `Outofcpu`
- allowing for lack of an exit code on a completed pod

## Why are we making these changes?

We've seen this situation in the wild, and we want to handle it cleanly in Cook.

## Example

<details>
  <summary>Click to expand</summary>

```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2019-11-22T05:17:45Z"
  labels:
    twosigma.com/cook-scheduler-job: gke-1
  name: 5dd76f79-5caf-44d4-a2b7-51a0b2980247
  namespace: cook
  resourceVersion: "2995861"
  selfLink: /api/v1/namespaces/cook/pods/5dd76f79-5caf-44d4-a2b7-51a0b2980247
  uid: 6ad800e4-0ce7-11ea-a908-42010a80018e
spec:
  containers:
  - command:
    - /bin/sh
    - -c
    - sleep 66666
    env:
    - name: COOK_INSTANCE_NUM
      value: "0"
    - name: COOK_INSTANCE_UUID
      value: 5dd76f79-5caf-44d4-a2b7-51a0b2980247
    - name: COOK_JOB_UUID
      value: 19112123-a109-45fc-9f4e-c7a3739aba8e
    - name: COOK_JOB_CPUS
      value: "0.4"
    - name: COOK_JOB_MEM_MB
      value: "512.0"
    - name: HOME
      value: /mnt/sandbox
    - name: MESOS_SANDBOX
      value: /mnt/sandbox
    image: gcr.io/google-containers/alpine-with-bash:1.0
    imagePullPolicy: IfNotPresent
    name: required-cook-job-container
    resources:
      limits:
        cpu: 400m
        memory: 512M
      requests:
        cpu: 400m
        memory: 512M
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /mnt/sandbox
      name: cook-workdir
    - mountPath: /mnt/cook-integration
      name: 8f272451-ad6c-4303-8419-84e8a345bc08
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-sqft6
      readOnly: true
    workingDir: /mnt/sandbox
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: gke-dpo-test-cluster-201-default-pool-8fdb29b8-t7bd
  priority: 0
  restartPolicy: Never
  schedulerName: default-scheduler
  securityContext:
    runAsUser: 0
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir: {}
    name: cook-workdir
  - hostPath:
      path: /tmp/cook-integration-mount
      type: DirectoryOrCreate
    name: 8f272451-ad6c-4303-8419-84e8a345bc08
  - name: default-token-sqft6
    secret:
      defaultMode: 420
      secretName: default-token-sqft6
status:
  message: 'Pod Node didn''t have enough resource: cpu, requested: 400, used: 921,
    capacity: 940'
  phase: Failed
  reason: OutOfcpu
  startTime: "2019-11-22T05:17:45Z"
```

</details>

### CODE REVIEW

**Comment by scrosby** [comment]
Do we know why the pod status is OutOfCPU? This error "shouldn't happen". I'm concerned that this papers over a severe bug. At the minimum, we should track how often these occur and make sure we followup as to why they occur.

**Comment by dposada** [comment]
Nope, not sure.

Good call, will add a log line for this.

File: scheduler/src/cook/kubernetes/controller.clj [hunk 1]
```diff
@@ -27,8 +27,10 @@
 (defn existing-state-equivalent?
   "Is the old and new state equivalent?"
   [old-state new-state]
-  (= old-state new-state) ; TODO. This is not right. We need to tune/tweak this to suppress otherwise identical states so we don't spam.
-  )
+  ; TODO:
+  ; This is not right. We need to tune/tweak this to
+  ; suppress otherwise identical states so we don't spam.
+  (= (dissoc old-state :ancillary) (dissoc new-state :ancillary)))
```
Comment (nsinkov) [inline scheduler/src/cook/kubernetes/controller.clj]: we decided we don't want the ancillary field

Comment (dposada) [reply scheduler/src/cook/kubernetes/controller.clj]: Removed, thanks

**Pull request merged**